File: src/backend/IAW.Vdf.Authoring/DryRun/DryRunPreviewer.cs
using IAW.Vdf.Abstractions.Evaluation;
using IAW.Vdf.Abstractions.Facts;
using IAW.Vdf.Abstractions.Outcomes;
using IAW.Vdf.Abstractions.ReferenceData;
using IAW.Vdf.Abstractions.Rules;
using IAW.Vdf.Abstractions.Triggers;
using IAW.Vdf.Core.Engine;
using IAW.Vdf.Core.Outcomes;
using IAW.Vdf.Core.Repositories;
using IAW.Vdf.Core.Time;

namespace IAW.Vdf.Authoring.DryRun;

/// <summary>Represents the outcome of running a candidate rule against a single fixture.</summary>
/// <param name="FixtureName">The name of the fixture file or fixture label.</param>
/// <param name="Applied">Whether the rule's AppliesWhen guard held (rule was evaluated).</param>
/// <param name="Produced">The type of the outcome produced, if the rule applied.</param>
/// <param name="Reason">The reason string from the produced outcome, if any.</param>
public sealed record DryRunHit(string FixtureName, bool Applied, OutcomeType? Produced, string? Reason);

/// <summary>Summary results for a dry-run preview over a set of fixtures.</summary>
/// <param name="Evaluated">The total number of fixtures evaluated.</param>
/// <param name="Hits">Per-fixture evaluation results.</param>
public sealed record DryRunResult(int Evaluated, IReadOnlyList<DryRunHit> Hits);

/// <summary>
/// Evaluates a candidate <see cref="RuleDefinition"/> against fact fixtures in a no-side-effects
/// sandbox, using a <see cref="CollectingOutcomeHandler"/> so no external state is mutated.
/// </summary>
public sealed class DryRunPreviewer
{
    private readonly IReferenceDataProvider _references;

    /// <summary>
    /// Creates a new <see cref="DryRunPreviewer"/> backed by the supplied reference-data provider.
    /// </summary>
    /// <param name="references">Reference data needed to resolve rule conditions.</param>
    public DryRunPreviewer(IReferenceDataProvider references)
    {
        _references = references;
    }

    /// <summary>
    /// Runs the candidate rule against each named fixture and returns a summary of hits.
    /// No external state is modified; the engine operates on clones of each <see cref="FactDocument"/>.
    /// </summary>
    /// <param name="candidate">The rule to preview.</param>
    /// <param name="fixtures">Named fact documents to evaluate.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>A <see cref="DryRunResult"/> containing per-fixture hit records.</returns>
    public async Task<DryRunResult> PreviewAsync(
        RuleDefinition candidate,
        IEnumerable<(string Name, FactDocument Facts)> fixtures,
        CancellationToken cancellationToken = default)
    {
        var handler = new CollectingOutcomeHandler();
        var engine = BuildEngine(candidate, handler);
        var hits = new List<DryRunHit>();

        foreach (var (name, facts) in fixtures)
        {
            cancellationToken.ThrowIfCancellationRequested();

            var request = new EvaluationRequest
            {
                Trigger = Trigger.OrderEvent("dry-run"),
                Facts = facts,
                AsOf = DateTimeOffset.UtcNow,
                RuleSet = null
            };

            var result = await engine.EvaluateAsync(request, cancellationToken).ConfigureAwait(false);

            // Find the trace for this specific rule.
            var trace = result.Trace.FirstOrDefault(t => t.RuleKey == candidate.Key);
            if (trace is null)
            {
                hits.Add(new DryRunHit(name, false, null, null));
                continue;
            }

            var produced = trace.Produced;
            hits.Add(new DryRunHit(name, trace.Applied, produced?.Type, produced?.Reason));
        }

        return new DryRunResult(hits.Count, hits);
    }

    /// <summary>
    /// Loads all <c>*.json</c> files from the specified directory as fixtures and previews the candidate rule.
    /// Files named <c>reference-data.json</c> are skipped.
    /// </summary>
    /// <param name="candidate">The rule to preview.</param>
    /// <param name="fixturesDirectory">Path to a directory containing fixture JSON files.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>A <see cref="DryRunResult"/> containing per-fixture hit records.</returns>
    public async Task<DryRunResult> PreviewFromDirectoryAsync(
        RuleDefinition candidate,
        string fixturesDirectory,
        CancellationToken cancellationToken = default)
    {
        var files = Directory
            .GetFiles(fixturesDirectory, "*.json")
            .Where(f => !string.Equals(
                Path.GetFileName(f),
                "reference-data.json",
                StringComparison.OrdinalIgnoreCase))
            .OrderBy(f => f, StringComparer.Ordinal)
            .ToList();

        var fixtures = files.Select(path =>
        {
            var json = File.ReadAllText(path);
            var facts = FactDocument.Parse(json);
            var name = Path.GetFileNameWithoutExtension(path);
            return (name, facts);
        });

        return await PreviewAsync(candidate, fixtures, cancellationToken).ConfigureAwait(false);
    }

    /// <summary>
    /// Locates the repository's <c>fixtures/</c> directory by walking upward from
    /// <see cref="AppContext.BaseDirectory"/>, then previews the candidate rule against all fixtures.
    /// </summary>
    /// <param name="candidate">The rule to preview.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>A <see cref="DryRunResult"/> containing per-fixture hit records.</returns>
    public Task<DryRunResult> PreviewFromRepoFixturesAsync(
        RuleDefinition candidate,
        CancellationToken cancellationToken = default)
    {
        var dir = FindDir("fixtures");
        return PreviewFromDirectoryAsync(candidate, dir, cancellationToken);
    }

    // ── Private helpers ──────────────────────────────────────────────────────────────────────────

    private VdfEngine BuildEngine(RuleDefinition candidate, CollectingOutcomeHandler handler)
    {
        var repository = new InMemoryRuleRepository(new[] { candidate });
        var clock = new FixedClock(DateTimeOffset.UtcNow);
        var selector = new RuleSelector();
        return new VdfEngine(repository, _references, selector, clock, new[] { handler });
    }

    private static string FindDir(string name)
    {
        var dir = AppContext.BaseDirectory;
        while (dir is not null)
        {
            var candidate = Path.Combine(dir, name);
            if (Directory.Exists(candidate))
                return candidate;
            dir = Directory.GetParent(dir)?.FullName;
        }
        var abs = $"/Users/bharath/Desktop/NeoGenomics/IAW/{name}";
        if (Directory.Exists(abs)) return abs;
        throw new DirectoryNotFoundException($"Could not locate '{name}' directory.");
    }
}

File: src/server/src/authoring/dry-run-previewer.ts
/**
 * Read-only dry-run preview of a CANDIDATE rule over a fixtures corpus.
 *
 * A faithful port of {@link ../../backend/IAW.Vdf.Authoring/DryRun/DryRunPreviewer.cs}:
 * the candidate is the ONLY rule in a throwaway engine, every produced outcome is
 * routed to a collecting (no-op) handler, and the engine already clones the facts —
 * so previewing has zero side effects on persisted state or the caller's input.
 * Each fixture yields a hit: did the rule apply, what did it produce, and why.
 */

import * as fs from 'fs';
import * as path from 'path';

import { FixedClock } from '../vdf/clock';
import { VdfEngine, OutcomeHandler } from '../vdf/engine';
import { ReferenceDataProvider } from '../vdf/reference-data';
import { JsonObject, Outcome, OutcomeType, RuleDefinition } from '../vdf/types';

/** The fixed instant a preview runs at (the corpus parity oracle's "now"). */
const PREVIEW_NOW = '2026-06-17T12:00:00+00:00';

/** The result of running the candidate rule against a single fixture. */
export interface DryRunHit {
  /** The fixture file name (without extension). */
  fixtureName: string;
  /** Whether the rule's AppliesWhen guard held (the rule was evaluated). */
  applied: boolean;
  /** The outcome type produced, if the rule produced one. */
  produced: OutcomeType | null;
  /** The reason string from the produced outcome, if any. */
  reason: string | null;
}

/** Summary of a dry-run preview over a fixtures corpus. */
export interface DryRunResult {
  /** Total fixtures evaluated. */
  evaluated: number;
  /** Per-fixture hit records. */
  hits: DryRunHit[];
}

/**
 * A no-op {@link OutcomeHandler} that records nothing externally — it exists only
 * to satisfy the engine's dispatch loop without mutating any state, guaranteeing
 * the preview is side-effect free.
 */
class CollectingOutcomeHandler implements OutcomeHandler {
  canHandle(): boolean {
    return true;
  }
  handle(): void {
    // Intentionally empty: a dry run produces no real-world effects.
  }
}

/** Evaluates a candidate rule against fact fixtures in a no-side-effects sandbox. */
export class DryRunPreviewer {
  constructor(private readonly references: ReferenceDataProvider) {}

  /**
   * Runs the candidate against each named fixture and returns per-fixture hits.
   * The engine clones each fact document, so inputs are never mutated.
   */
  preview(
    candidate: RuleDefinition,
    fixtures: ReadonlyArray<{ name: string; facts: JsonObject }>,
  ): DryRunResult {
    const engine = new VdfEngine(
      [candidate],
      this.references,
      new FixedClock(PREVIEW_NOW),
      [new CollectingOutcomeHandler()],
    );

    const hits: DryRunHit[] = fixtures.map(({ name, facts }) => {
      const result = engine.evaluate({ facts, asOf: PREVIEW_NOW });
      const trace = result.trace.find((t) => t.ruleKey === candidate.key);
      if (trace === undefined) {
        return {
          fixtureName: name,
          applied: false,
          produced: null,
          reason: null,
        };
      }
      const produced: Outcome | null = trace.produced;
      return {
        fixtureName: name,
        applied: trace.applied,
        produced: produced?.type ?? null,
        reason: produced?.reason ?? null,
      };
    });

    return { evaluated: hits.length, hits };
  }

  /**
   * Loads every `*.json` fixture from a directory (skipping `reference-data.json`)
   * and previews the candidate against them, ordered deterministically by name.
   */
  previewFromDirectory(
    candidate: RuleDefinition,
    fixturesDirectory: string,
  ): DryRunResult {
    const fixtures = fs
      .readdirSync(fixturesDirectory)
      .filter(
        (f) => f.endsWith('.json') && f.toLowerCase() !== 'reference-data.json',
      )
      .sort((a, b) => a.localeCompare(b))
      .map((file) => {
        const raw = fs.readFileSync(path.join(fixturesDirectory, file), 'utf8');
        return {
          name: path.basename(file, '.json'),
          facts: JSON.parse(raw) as JsonObject,
        };
      });

    return this.preview(candidate, fixtures);
  }

  /**
   * Locates the repository's `fixtures/` directory by walking up from this module
   * and previews the candidate against all fixtures.
   */
  previewFromRepoFixtures(candidate: RuleDefinition): DryRunResult {
    return this.previewFromDirectory(candidate, findFixturesDir());
  }
}

/** Walks up from this module to find the repo-root `fixtures/` directory. */
function findFixturesDir(): string {
  let dir: string | undefined = __dirname;
  while (dir !== undefined) {
    const candidate = path.join(dir, 'fixtures');
    if (fs.existsSync(candidate) && fs.statSync(candidate).isDirectory()) {
      return candidate;
    }
    const parent = path.dirname(dir);
    dir = parent === dir ? undefined : parent;
  }
  const abs = '/Users/bharath/Desktop/NeoGenomics/IAW/fixtures';
  if (fs.existsSync(abs)) {
    return abs;
  }
  throw new Error("Could not locate the 'fixtures' directory.");
}
